docs/tutorials/tutorial_2.rst

```diff
@@ -0,0 +1,537 @@
+            return numNodesVisited == N;
+        }
+    };
+
+----
+
+Writing test spec
+-----------------
+
+The first difference in the ``TestSpec`` class compared to the first tutorial is in the sample test cases declaration. We need to declare which subtasks will have this sample case as one of their testcases. For the sample input in the problem above, since M ≤ 20 and not all W are equal, then the sample input will be included in subtask 1 and subtask 3.
+
+``minimum-spanning-tree/spec.cpp`` (continued):
+
+.. sourcecode:: cpp
+
+    class TestSpec : public BaseTestSpec<ProblemSpec> {
+    protected:
+        void SampleTestCase1() {
+            Subtasks({1, 3});
+            Input({
+                "3 3",
+                "0 1 2",
+                "1 2 3",
+                "2 0 4"
+            });
+            Output({
+                "5"
+            });
+        }
+    };
+
+The ``BeforeTestCase()`` has the same syntax compared to the previous tutorial. However, here we need to clear more vectors before every test case.
+
+.. sourcecode:: cpp
+
+    void BeforeTestCase() {
+        U.clear();
+        V.clear();
+        W.clear();
+    }
+
+Before creating the actual test cases, let us create the private helper methods that will help us in creating the actual test cases later. Similar to the private helper methods in ``ProblemSpec``, these methods are very modular and we can always reuse them for another problem.
+
+.. sourcecode:: cpp
+
+    void randomWeight(int M, vector<int> &W, int minW = 1, int maxW = 1000) {
+        for (int i = 0; i < M; ++i) {
+            W.push_back(rnd.nextInt(minW, maxW));
+        }
+    }
+
+    void renumber(int N, vector<int> &U, vector<int> &V) {
+        vector<int> permutation;
+        for (int i = 0; i < N; ++i) {
+            permutation.push_back(i);
+        }
+        rnd.shuffle(permutation.begin(), permutation.end());
+        for (int i = 0; i < (int)U.size(); ++i) {
+            U[i] = permutation[U[i]];
+            V[i] = permutation[V[i]];
+        }
+    }
+
+    void randomTree(int N, vector<int> &U, vector<int> &V) {
+        for (int i = 1; i < N; ++i) {
+            U.push_back(i);
+            V.push_back(rnd.nextInt(0, i - 1));
+        }
+        renumber(N, U, V);
+    }
+
+    void randomGraph(int N, int M, vector<int> &U, vector<int> &V) {
+        randomTree(N, U, V);
+        while (U.size() < M) {
+            int u = rnd.nextInt(0, N - 2);
+            int v = rnd.nextInt(u + 1, N - 1);
+            U.push_back(u);
+            V.push_back(v);
+        }
+    }
+
+Now, let us move on to the creating test cases itself. The actual test cases for a problem involving test groups are challenging. The first step is to draw the venn diagram of the subtasks.
+
+The properties of the subtasks of the above problem are :
+
+1. All testcases inside subtask 1 and subtask 2 are also inside subtask 3, and subtask 3 have more testcases that is neither in subtask 1 nor subtask 2.
```

Comment:
- inside -> in (everywhere)
- subtask 3 have more testcases that is -> subtask 3 contains some other test cases that are